content/blog/2019/knative-activator-adapter/index.md

```diff
@@ -0,0 +1,93 @@
+than Envoy is.
+
+It is an attribute processing engine that uses operator-supplied configuration to map request attributes from the proxy into calls
+to these backend systems via a pluggable set of adapters. Adapters enable **Mixer** to expose a single consistent API, independent of the
+infrastructure backends in use. The exact set of adapters used at runtime is determined through operator configuration and can easily
+be extended to target new or custom infrastructure backends.
+
+Here we utilized these benefits to implement the Knative scale-from-zero logic in a special Mixer
+[out-of-process adapter](https://github.com/istio/istio/wiki/Mixer-Out-Of-Process-Adapter-Dev-Guide)
+which allows you to use any programming language and to build and maintain your extension as a stand-alone
+program without the need (as in previous Istio releases) to build the whole Istio proxy.
+The diagram below presents the Knative setup for scale-from-zero using our adapter.
+
+{{< image width="60%" link="knative-mixer-adapter.png" caption="Knative scale-from-zero" >}}
+
+In this setup, there is no need to change the routing from/to **Activator** for an idle application as in the original Knative setup.
+When the Istio proxy, inside an ingress gateway, receives a new request for an idle application, it informs **Mixer**, including all the
+relevant metadata information.
+**Mixer** then calls our adapter which triggers the Knative **Autoscaler** using the original Knative protocol.
+Note that we do not need to deal with buffering, retries and load-balancing because it is already handled by the Istio proxy.
+
+This Istio capability of Mixer's extension makes it very simple to convert a complex networking-based application logic into a few lines
+of code of a Mixer adapter.
+The actual code of the [adapter](https://github.com/zachidan/istio-kactivator) is presented below.
+
+```go
+// Send a StatMessage to the Knative AutoScaler over the WebSocket using the Knative original protocol
+func (s *GrpcAdapter) sendStatMessage(r *authorization.HandleAuthorizationRequest) {
+    namespace := r.Instance.Action.Namespace
+    service := r.Instance.Action.Service
+    sm := autoscaler.StatMessage{Key: fmt.Sprintf("%s/%s", namespace, service)}
+    sm.Stat.PodName = s.podName
+    sm.Stat.AverageConcurrentRequests = 1.0
+    sm.Stat.RequestCount = 1
+    s.statSink.Send(&sm)
+}
+```
+
+When the adapter receives a message from **Mixer**, it sends a `StatMessage` directly to **Autoscaler**
+component using the Knative protocol.
+The metadata information (`namespace` and `service name`) required by **Autoscaler** are transferred by Istio proxy to
+**Mixer** and from there to the adapter.
+
+## Summary
+
+We ran a performance benchmark and compared the cold-start time of the original Knative setup vs. the Istio Mixer adapter setup.
+The results show similar cold-start time. However, the complexity of the code using Mixer adapter is much smaller
```

Comment:
time/times.

Reply:
However, the complexiy...  The implementation using the Mixer adapter has greater simplicity.  It is not necessary to handle low-level network-based mechanisms as these are handled by (Envoy?).